core/src/listener/listener.ts

```diff
@@ -220,36 +199,50 @@ function latestJob({
       observedBlock = Math.max(0, latestBlock - 1)
     }
 
-    const logPrefix = generateListenerLogPrefix(contractAddress, observedBlock, latestBlock)
     try {
       if (latestBlock > observedBlock) {
-        await redisClient.set(observedBlockRedisKey, latestBlock)
-
         // The `observedBlock` block number is already processed,
         // therefore we do not need to re-query the same event in such
         // block again.
-        const events = await state.queryEvent(contractAddress, observedBlock + 1, latestBlock)
-        for (const [index, event] of events.entries()) {
-          const outData: IProcessEventListenerJob = {
-            contractAddress,
-            event
+        const lockObservedBlock = observedBlock + 1
+        for (let blockNumber = lockObservedBlock; blockNumber <= latestBlock; ++blockNumber) {
+          const events = await state.queryEvent(contractAddress, blockNumber, blockNumber)
+          for (const [_, event] of events.entries()) {
+            const jobMetadata = await processFn(event)
+            if (jobMetadata) {
+              const { jobId, jobName, jobData, jobQueueSettings } = jobMetadata
+              const queueSettings = jobQueueSettings ? jobQueueSettings : LISTENER_JOB_SETTINGS
+              await workerQueue.add(jobName, jobData, {
+                jobId,
+                ...queueSettings
+              })
+              logger.debug(`Listener submitted job [${jobId}] for [${jobName}]`)
+            }
           }
-          const jobId = getUniqueEventIdentifier(event, index)
-          await processEventQueue.add('latest', outData, {
-            jobId,
-            ...LISTENER_JOB_SETTINGS
+          await upsertListenerObservedBlock({
+            blockKey: observedBlockRedisKey,
+            blockNumber,
+            logger: this.logger
```

Comment:
I have a suggestion, I might be missing something so feel free to point out if I'm wrong

I was thinking that `nestjs` [lifecycle hook](https://docs.nestjs.com/fundamentals/lifecycle-events) could be utilized, I'm assuming you didn't fully remove redis implementation to store and load block numbers

So maybe storing block number into database can happen only on service shutdown

meanwhile service is running redis can be utilized